### PR TITLE
heapster: preserve GID when running in Docker

### PR DESCRIPTION
**Pull request by ghost**
Following up https://github.com/kubernetes/heapster/issues/1658#issuecomment-331159660 , this PR intends to bring the default Docker behavior in order to preserve the relevant GID set to ``65534`` (nobody) since by default Kubernetes would not respect the image's ``USER uid:gid`` set in the Dockerfile. Not sure whether there is an opened issue/pr for the Kubernetes to address this.

----

Prepare a Dockerfile with the user 65534:65534 (nobody:nobody)

```
FROM scratch
# wget https://busybox.net/downloads/binaries/1.27.1-i686/busybox
# chmod +x busybox
COPY busybox /busybox
USER 65534:65534
ENTRYPOINT ["/busybox"]
CMD ["id"]
```

Build the image

```
docker build -t scratch .
```

Running the image shows uid:gid as expected - 65534:65534.

```
docker run --rm -ti scratch
uid=65534 gid=65534
```

When uid is explicitly set to 65534, then gid is not getting picked, because ``scratch`` image does not have ``/etc/passwd`` file.

```
docker run --rm -u 65534 -ti scratch
uid=65534 gid=0
```

```
docker run --rm -u nobody  -ti scratch
docker: Error response from daemon: linux spec user: unable to find user nobody: no matching entries in passwd file.
```

```
docker run --rm -ti scratch cat /etc/passwd
cat: can't open '/etc/passwd': No such file or directory
```

As a remedy, let's add the ``/etc/passwd`` file to the ``scratch`` image

```
echo 'nobody:x:65534:65534:nobody:/home:/bin/false' > passwd
```

Dockerfile:

```
FROM scratch
# wget https://busybox.net/downloads/binaries/1.27.1-i686/busybox
# chmod +x busybox
COPY busybox /busybox
COPY passwd /etc/passwd
USER 65534:65534
ENTRYPOINT ["/busybox"]
CMD ["id"]
```

The gid is getting picked now:

```
docker run --rm -ti -u nobody scratch 
uid=65534(nobody) gid=65534
```

```
docker run --rm -ti -u 65534 scratch 
uid=65534(nobody) gid=65534
```

### building and pushing the images

```
docker build -t andrey01/scratch:passwd .
docker build -t andrey01/scratch:nopasswd .
docker push andrey01/scratch:passwd
docker push andrey01/scratch:nopasswd
```

### Preparing the jobs

k8s-job-scratch-passwd.yml:

```
kind: Job
apiVersion: batch/v1
metadata:
  name: scratch-passwd
spec:
  template:
    metadata:
      labels:
        app: scratch-passwd
    spec:
      restartPolicy: OnFailure
      containers:
      - name: scratch-passwd
        image: andrey01/scratch:passwd
        imagePullPolicy: Always
```

k8s-job-scratch-nopasswd.yml:

```
kind: Job
apiVersion: batch/v1
metadata:
  name: scratch-nopasswd
spec:
  template:
    metadata:
      labels:
        app: scratch-nopasswd
    spec:
      restartPolicy: OnFailure
      containers:
      - name: scratch-nopasswd
        image: andrey01/scratch:nopasswd
        imagePullPolicy: Always
```

### Running the jobs

```
kubectl create -f k8s-job-scratch-nopasswd.yml 
kubectl create -f k8s-job-scratch-passwd.yml 
```

## Checking the output

```
$ kubectl get jobs
NAME               DESIRED   SUCCESSFUL   AGE
scratch-nopasswd   1         1            19s
scratch-passwd     1         1            16s
```

```
$ kubectl logs $(kubectl get pods --show-all -l app=scratch-passwd --output jsonpath='{.items[0].metadata.name}')
uid=65534(nobody) gid=65534
```

```
$ kubectl logs $(kubectl get pods --show-all -l app=scratch-nopasswd --output jsonpath='{.items[0].metadata.name}')
uid=65534 gid=0
```

As you can see gid is 0 for the scratch container which does not have ``/etc/passwd`` file containing the relevant uid:gid pair.

## Cleanup

```
kubectl delete job scratch-nopasswd scratch-passwd
```

### CODE REVIEW

**Comment by arno01** [comment]
Humm, Google wants me to sign the Contributor License Agreement (CLA), but I do not have the account at Google. Is there any workaround how I can sign the CLA ?

File: deploy/docker/Dockerfile [hunk 1]
```diff
@@ -3,6 +3,9 @@ FROM scratch
 COPY heapster eventer /
 COPY ca-certificates.crt /etc/ssl/certs/
 
+# Make sure a GID for a user nobody (65534) is preserved when running in K8s.
```
Comment (andyxning) [inline deploy/docker/Dockerfile]: > running in K8s.

This is also required in Docker not only in Kubernetes, IMO.

File: deploy/docker/passwd [hunk 2]
```diff
@@ -0,0 +1 @@
+nobody:x:65534:65534:nobody:/home:/bin/false
```
Comment (andyxning) [inline deploy/docker/passwd]: `nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin` is more preferable. `nobody` should not be able to login.

**Comment by andyxning** [comment]
@arno01 I am not quite familiar with the CLA and Google account. But i am afraid you may need to apply for one Google account to accomplish signing the CLA.

As for adding `nobody` user and group with UID and GID `65534`, an easier way is to use `alpine:3.6` which is smallest linux distribution for image.

@DirectXMan12 @piosz WDYT.

**Comment by arno01** [comment]
I signed it!

**Comment by arno01** [comment]
@andyxning alpine:3.6 sounds reasonable to me as well.

**Comment by andyxning** [comment]
@arno01 `alpine` is small enough and should not add too much overhead to heapster image. At the same time, heapster will get the full linux environments.

Could you please adjust the PR with changing `scratch` to `alpine:3.6`?

**Comment by arno01** [comment]
@andyxning done.

How do you think, should we add some sort of a test which would be always ensuring the expected UID:GID pair is present in the ``/etc/passwd`` file?

E.g. something like:

```
RUN grep -q "65534:65534" /etc/passwd
```

**Comment by andyxning** [comment]
@arno01 

> should we add some sort of a test which would be always ensuring the expected UID:GID pair is present in the /etc/passwd file?

Sure, this is definitely good. In fact i am not sure whether it is easy to add such a test.

> about doing `RUN apk --update add ca-certificates`

Agreed. @arno01 maybe this should be 
```
RUN apk update
RUN apk --no-cache add ca-certificates
```

**Comment by piosz** [comment]
@timstclair @crassirostris any thoughs on this?

**Comment by piosz** [comment]
cc @luxas

**Review by luxas — APPROVED** [review]
Hi @arno01, thanks for this contribution :wave:!

It must be multiarch ready before we can merge this, right now this would only work for amd64

Please look at the other Dockerfiles/Makefiles for reference how to do it.

**Pull request closed without merging**